mypy/plugins/singledispatch.py

```diff
@@ -0,0 +1,178 @@
+            }  # type: RegisterCallableInfo
+            register_callable.type.metadata[METADATA_KEY] = register_metadata  # type: ignore
+            return register_callable
+        elif isinstance(first_arg_type, CallableType):
+            # TODO: do more checking for registered functions
+            register_function(ctx.type, first_arg_type)
+
+    # register doesn't modify the function it's used on
+    return ctx.default_return_type
+
+
+def register_function(singledispatch_obj: Instance, func: Type,
+                      register_arg: Optional[Type] = None) -> None:
+
+    func = get_proper_type(func)
+    if not isinstance(func, CallableType):
+        return
+    metadata = get_singledispatch_info(singledispatch_obj)
+    dispatch_type = get_dispatch_type(func, register_arg)
+    if dispatch_type is None:
+        # TODO: report an error here that singledispatch requires at least one argument
+        # (might want to do the error reporting in get_dispatch_type)
+        return
+    # TODO: report an error if we're overwriting another function (which would happen if multiple
+    # registered functions have the same dispatch type)
+    metadata['registered'][dispatch_type] = func
+
+
+def get_dispatch_type(func: CallableType, register_arg: Optional[Type]) -> Optional[Type]:
+    if register_arg is not None:
+        return register_arg
+    if func.arg_types:
+        return func.arg_types[0]
+    return None
+
+
+def call_singledispatch_function_after_register_argument(ctx: MethodContext) -> Type:
+    """Called on the function after passing a type to register"""
+    register_callable = ctx.type
+    if isinstance(register_callable, Instance):
+        metadata = cast(RegisterCallableInfo, register_callable.type.metadata[METADATA_KEY])
+        func = get_first_arg(ctx.arg_types)
+        if func is not None:
+            register_arg = metadata['register_type']
+            singledispatch_obj = metadata['singledispatch_obj']
+            register_function(singledispatch_obj, func, register_arg)
+    return ctx.default_return_type
+
+
+def rename_func(func: CallableType, new_name: CallableType) -> CallableType:
+    """Return a new CallableType that is `function` with the name of `new_name`"""
+    if new_name.name is not None:
+        signature_used = func.with_name(new_name.name)
+    else:
+        signature_used = func
+    return signature_used
+
+
+def call_singledispatch_function_callback(ctx: MethodSigContext) -> CallableType:
+    if not isinstance(ctx.type, Instance):
+        return ctx.default_signature
+    metadata = get_singledispatch_info(ctx.type)
+    first_arg = get_first_arg(ctx.args)
+    if first_arg is None:
+        return ctx.default_signature
+    # TODO: find a way to get the type of the first argument with the public API
+    # (expr_checker probably isn't part of the public API)
+    passed_type = cast(TypeChecker, ctx.api).expr_checker.accept(first_arg)
```

Comment:
The issue here is that we don't have access to the types because this is a `method_signature_hook` and not a `method_hook`, but we need to be a `method_signature_hook` because the whole point is the argument typechecking?

We could maybe add a type check method to `CheckerPluginInterface` but I'd want to think about how generally useful it would be. I don't really object to this cast.